### PR TITLE
Add support for zone anti_affinity, refactor affinity

### PR DESCRIPTION
**Pull request by addisonj**
This commit has two major changes:
1. It refactors affinity support to use a partial template to centralize
   and simplify the affinity calculation
2. It adds support for zone affinity, this is useful for cases where a
   user wants to ensure a cluster is spread across zones

*(If this PR fixes a github issue, please add `Fixes #<xyz>`.)*

Fixes #<xyz>

*(or if this PR is one task of a github issue, please add `Master Issue: #<xyz>` to link to the master issue.)*

Master Issue: #<xyz>

### Motivation

*Explain here the context, and why you're making that change. What is the problem you're trying to solve.*

### Modifications

*Describe the modifications you've done.*

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (10MB)*
  - *Extended integration test for recovery after broker failure*

### Documentation

Check the box below.

Need to update docs? 

- [ ] `doc-required` 
  
  (If you need help on updating docs, create a doc issue)
  
- [ ] `no-need-doc` 
  
  (Please explain why)
  
- [ ] `doc` 
  
  (If this PR contains doc changes)

### CODE REVIEW

File: charts/sn-platform/templates/bookkeeper/bookkeeper-cluster.yaml [hunk 1]
```diff
@@ -61,48 +61,7 @@ spec:
...
-                    values:
-                      - {{ .Release.Name }}
-                  - key: "component"
-                    operator: In
-                    values:
-                      - {{ .Values.bookkeeper.component }}
-              topologyKey: "kubernetes.io/hostname"
-          {{ end }}
-      {{- end }}
+      {{- include "pulsar.antiAffinityRules" (dict "Values" .Values "thisAffinity" .Values.bookkeeper.affinity "Component" .Values.bookkeeper.component "Relase" .Release) }}
```
Comment (MarvinCai) [inline charts/sn-platform/templates/bookkeeper/bookkeeper-cluster.yaml:64]:
```suggestion
      {{- include "pulsar.antiAffinityRules" (dict "Values" .Values "thisAffinity" .Values.bookkeeper.affinity "Component" .Values.bookkeeper.component "Release" .Release  "Chart" .Chart) | indent 6 }}
```

Comment (MarvinCai) [reply charts/sn-platform/templates/bookkeeper/bookkeeper-cluster.yaml:64]: same for all other includes

File: charts/sn-platform/templates/_helpers.tpl [hunk 2]
```diff
@@ -157,4 +157,92 @@ jvmOptions:
...
+            values:
+              - {{ .Release.Name }}
+          - key: "component"
+            operator: In
+            values:
+              - {{ .Component }}
+      topologyKey: "kubernetes.io/hostname"
+      {{ if and .Values.affinity.zone_anti_affinity .thisAffinity.zone_anti_affinity }}
+  preferredDuringSchedulingIgnoredDuringExecution:
+    - weight: {{ .thisAffinity.zone_anti_affinity_weight | .Values.affinity.zone_anti_affinity_weight | default "100" }}
```
Comment (MarvinCai) [inline charts/sn-platform/templates/_helpers.tpl:188]: here the .Values.affinity.zone_anti_affinity_weight will be treated as a function and will throw error, seems we can't use such way to choose value based on priority, I think it has to be either
`{{ .thisAffinity.zone_anti_affinity_weight | default .Values.affinity.zone_anti_affinity_weight }}
`or 
`{{ .thisAffinity.zone_anti_affinity_weight | default "100" }}`

File: charts/sn-platform/templates/_helpers.tpl [hunk 3]
```diff
@@ -157,4 +157,92 @@ jvmOptions:
...
   {{- end }}
-{{- end }}
\ No newline at end of file
+{{- end }}
+
+{{/*
+AntiAffinity Rules
+*/}}
+{{- define "pulsar.antiAffinityRules" }}
+{{- if .thisAffinity.affinity.customRules }}
```
[the remark this answers is not in the document]
Comment (MarvinCai) [reply charts/sn-platform/templates/_helpers.tpl:166]: need to be `.thisAffinity.customRules` as we are passing `"thisAffinity" .Values.bookkeeper.affinity`

**Review by MarvinCai — COMMENTED** [review]
the abstraction looks good, comments are based on test with `helm template`

**Comment by Huanli-Meng** [comment]
@addisonj @ericsyh do we need to update doc for this PR? Thanks.

**Pull request merged**